kube-runtime/src/utils/backoff_resettable.rs

```diff
@@ -0,0 +1,56 @@
+use std::{ops::DerefMut, time::Duration};
+
+use backon::{Backoff, BackoffBuilder};
+
+/// A [`Backoff`] that can also be reset.
+///
+/// Implemented by [`ResettableBackoffWrapper`].
+// Separated into a trait so that it can be used as a trait object, erasing the backing [`BackoffBuilder`].
+pub trait ResettableBackoff: Backoff {
+    fn reset(&mut self);
+}
+
+impl ResettableBackoff for Box<dyn ResettableBackoff + Send> {
+    /// Reset the [`Backoff`] to its initial state.
+    fn reset(&mut self) {
+        Box::deref_mut(self).reset();
+    }
+}
+
+/// Implements [`ResettableBackoff`] by reconstructing the backing [`Backoff`] each time [`Self::reset`] has been called.
+#[derive(Debug)]
+pub struct ResettableBackoffWrapper<B: BackoffBuilder> {
+    backoff_builder: B,
+    current_backoff: Option<B::Backoff>,
+}
+
+impl<B: BackoffBuilder> ResettableBackoffWrapper<B> {
+    pub fn new(backoff_builder: B) -> Self {
+        Self {
+            backoff_builder,
+            current_backoff: None,
+        }
+    }
+}
+
+impl<B: BackoffBuilder + Default> Default for ResettableBackoffWrapper<B> {
+    fn default() -> Self {
+        Self::new(B::default())
+    }
+}
```

Comment:
fwiw could probably be derived with `educe`, but the others probably not, so probably not worth doing proc macros for it

Reply:
Yeah for what it's worth I tried to do this with the default `Default` derive but that seemed to end up with the wrong trait bounds for reasons that I'm still not entirely clear on.